pkg/scheduler/queue/queue_test.go

```diff
@@ -18,102 +19,126 @@ import (
-	users := make([]string, 0, numTenants)
-	requests := make([]string, 0, numTenants)
-
-	for n := 0; n < b.N; n++ {
-		q := NewRequestQueue(maxOutstandingPerTenant, 0,
-			promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"user"}),
-			promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"user"}),
-			promauto.With(nil).NewHistogram(prometheus.HistogramOpts{}),
-		)
-
-		for ix := 0; ix < queriers; ix++ {
-			q.RegisterQuerierConnection(fmt.Sprintf("querier-%d", ix))
-		}
-
-		queues = append(queues, q)
-
-		for j := 0; j < numTenants; j++ {
-			requests = append(requests, fmt.Sprintf("%d-%d", n, j))
-			users = append(users, strconv.Itoa(j))
-		}
-	}
-
-	b.ResetTimer()
-	for n := 0; n < b.N; n++ {
-		for i := 0; i < maxOutstandingPerTenant; i++ {
-			for j := 0; j < numTenants; j++ {
-				err := queues[n].EnqueueRequest(users[j], requests[j], 0, nil)
-				if err != nil {
-					b.Fatal(err)
-				}
+func BenchmarkConcurrentQueueOperations(b *testing.B) {
+	req := "the query request"
+	maxQueriers := 0 // TODO: test with limit enforced?
+
+	for _, numTenants := range []int{1, 10, 1000} {
+		b.Run(fmt.Sprintf("%v tenants", numTenants), func(b *testing.B) {
+			for _, numProducers := range []int{10, 25} { // Query-frontends run 5 parallel streams per scheduler by default, and we typically see 2-5 frontends running at any one time.
+				b.Run(fmt.Sprintf("%v concurrent producers", numProducers), func(b *testing.B) {
+					for _, numConsumers := range []int{16, 160, 1600} { // Queriers run with parallelism of 8 when query sharding is enabled.
+						b.Run(fmt.Sprintf("%v concurrent consumers", numConsumers), func(b *testing.B) {
+							queueLength := promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"user"})
+							discardedRequests := promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"user"})
+							enqueueDuration := promauto.With(nil).NewHistogram(prometheus.HistogramOpts{})
+							queue := NewRequestQueue(100, 0, queueLength, discardedRequests, enqueueDuration)
+
+							start := make(chan struct{})
+							producersAndConsumers, ctx := errgroup.WithContext(context.Background())
+							require.NoError(b, queue.starting(ctx))
+							b.Cleanup(func() {
+								require.NoError(b, queue.stop(nil))
+							})
+
+							requestCount := func(total int, instanceCount int, instanceIdx int) int {
+								count := total / instanceCount
+								totalCountWithoutAdjustment := count * instanceCount
+
+								if totalCountWithoutAdjustment >= total {
+									return count
+								}
+
+								additionalRequests := total - totalCountWithoutAdjustment
+
+								if instanceIdx < additionalRequests {
+									// If we can't perfectly spread requests across all instances, assign remaining requests to the first instances.
+									return count + 1
+								}
+
+								return count
+							}
+
+							// TODO: limit the rate of production?
+							startProducer := func(producerIdx int) error {
```

Comment:
[nit] Start makes me thing it start a goroutine, but this is synchronous and the goroutine is fired by the caller. We may simply call it `runProducer`.

```suggestion
							runProducer := func(producerIdx int) error {
```

Reply:
Dimitar had the same suggestion, changed in de00c70.